espnet/nets/pytorch_backend/e2e_asr_maskctc.py

```diff
@@ -0,0 +1,212 @@
+# Copyright 2020 Johns Hopkins University (Shinji Watanabe)
+#                Waseda University (Yosuke Higuchi)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+"""
+Mask CTC based non-autoregressive speech recognition model (pytorch).
+See https://arxiv.org/abs/2005.08700 for the detail.
+
+"""
+import logging
+import math
+from itertools import groupby
+
+import numpy
+import torch
+
+from espnet.nets.pytorch_backend.e2e_asr_transformer import E2E as E2ETransformer
+from espnet.nets.pytorch_backend.e2e_asr import CTC_LOSS_THRESHOLD
+from espnet.nets.pytorch_backend.nets_utils import make_non_pad_mask
+from espnet.nets.pytorch_backend.nets_utils import th_accuracy
+from espnet.nets.pytorch_backend.maskctc.add_mask import mask_uniform
+
+
+class E2E(E2ETransformer):
+    """E2E module.
+
+    :param int idim: dimension of inputs
+    :param int odim: dimension of outputs
+    :param Namespace args: argument Namespace containing options
+
+    """
+
+    @staticmethod
+    def add_arguments(parser):
+        """Add arguments."""
+        E2ETransformer.add_arguments(parser)
+        E2E.add_maskctc_arguments(parser)
+
+        return parser
+
+    @staticmethod
+    def add_maskctc_arguments(parser):
+        """Add arguments for maskctc model."""
+        group = parser.add_argument_group("maskctc specific setting")
+
+        return parser
+
+    def __init__(self, idim, odim, args, ignore_id=-1):
+        """Construct an E2E object.
+
+        :param int idim: dimension of inputs
+        :param int odim: dimension of outputs
+        :param Namespace args: argument Namespace containing options
+        """
+        odim += 1 # for the mask token
+
+        super().__init__(idim, odim, args, ignore_id)
+        assert 0.0 <= self.mtlalpha < 1.0, "mtlalpha should be [0.0, 1.0)"
+
+        self.mask_token = odim - 1
+        self.sos = odim - 2
+        self.eos = odim - 2
+        self.odim = odim
+
+    def forward(self, xs_pad, ilens, ys_pad):
+        """E2E forward.
+
+        :param torch.Tensor xs_pad: batch of padded source sequences (B, Tmax, idim)
+        :param torch.Tensor ilens: batch of lengths of source sequences (B)
+        :param torch.Tensor ys_pad: batch of padded target sequences (B, Lmax)
+        :return: ctc loss value
+        :rtype: torch.Tensor
+        :return: attention loss value
+        :rtype: torch.Tensor
+        :return: accuracy in attention decoder
+        :rtype: float
+        """
+        # 1. forward encoder
+        xs_pad = xs_pad[:, : max(ilens)]  # for data parallel
+        src_mask = make_non_pad_mask(ilens.tolist()).to(xs_pad.device).unsqueeze(-2)
+        hs_pad, hs_mask = self.encoder(xs_pad, src_mask)
+        self.hs_pad = hs_pad
+
+        # 2. forward decoder
+        ys_in_pad, ys_out_pad = mask_uniform(
+            ys_pad, self.mask_token, self.eos, self.ignore_id
+        )
+        ys_mask = (ys_in_pad != self.ignore_id).unsqueeze(-2)
```

Comment:
> 

I think this may not be an error? `ys_mask` is used for masking pad-token in decoder's self-attention thus should be calculated from `ys_in_pad`. 
In fact, the `ys_mask` calculated using `ys_out_pad` is meaningful only when calculating accuracy or loss, this has been done inside function `th_accuracy` and forwardpass of  `self.criterion`.

Reply:
My point is that `ys_in_pad` is padded with `eos`. So the expected padding indices are not padded accutually.

Reply:
Thank you for the comments. This way of padding is not correct as hirofumi pointed out. I'll rerun the experiments to see if this bug has effects on mask ctc's performance.